File: 338-CountingBits/338-CountingBits.cpp
// Last updated: 6/6/2025, 3:32:11 PM
class Solution {
public:
    int popCount(int n) 
    {
        // Count number of bit flips by BK's Algorithm
        // This is known as Brian Kernighan's algorithm for counting set bits. Every iteration removes the lowest set bit from n, so the loop runs exactly popcount(n) times.
        int count;
        for (count = 0; n != 0; ++count) { n &= n - 1; } 
        return count;
    }

    vector<int> countBits(int n) 
    {
        // Get bit count and store in vector index position
        vector<int> ans(n + 1, 0);
        for (int x = 0; x <= n; ++x) { ans[x] = popCount(x); }
        return ans;
    }
};

File: 338-CountingBits/338-CountingBits.rs
// Last updated: 6/6/2025, 3:32:09 PM
impl Solution 
{
    pub fn count_bits(n: i32) -> Vec<i32> 
    {
        fn pop_count(mut n: i32) -> i32 
        {
            let mut count = 0;
            while n != 0 
            {
                // Clear the least significant 1-bit
                n &= n - 1; 
                count += 1;
            }
            count
        }

        let mut ans = Vec::with_capacity((n + 1) as usize);
        for x in 0..=n { ans.push(pop_count(x)); }
        ans
    }
}
